packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3614,9 +3643,149 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    }
+    return false;
+  }
+
+  void _handleContextMenuOnParentScroll(ScrollNotification notification) {
+    // Do some preliminary checks to avoid expensive subtree traversal.
+    if (notification is! ScrollStartNotification
+       && notification is! ScrollEndNotification) {
+      return;
+    }
+    if (notification is ScrollStartNotification
+       && _dataWhenToolbarShowScheduled != null) {
+      return;
+    }
+    if (notification is ScrollEndNotification
+       && _dataWhenToolbarShowScheduled == null) {
+      return;
+    }
+    if (notification is ScrollEndNotification
+       && _dataWhenToolbarShowScheduled!.value != _value) {
+      _dataWhenToolbarShowScheduled = null;
+      _disposeScrollNotificationObserver();
+      return;
+    }
+    if (_isInternalScrollableNotification(notification.context)) {
+      return;
+    }
+    if (!_scrollableNotificationIsFromSameSubtree(notification.context)) {
+      return;
+    }
+    _handleContextMenuOnScroll(notification);
+  }
+
+  Rect _calculateDeviceRect() {
+    final Size screenSize = MediaQuery.sizeOf(context);
+    final ui.FlutterView view = View.of(context);
+    final double obscuredVertical = (view.padding.top + view.padding.bottom + view.viewInsets.bottom) / view.devicePixelRatio;
+    final double obscuredHorizontal = (view.padding.left + view.padding.right) / view.devicePixelRatio;
+    final Size visibleScreenSize = Size(screenSize.width - obscuredHorizontal, screenSize.height - obscuredVertical);
+    return Rect.fromLTWH(view.padding.left / view.devicePixelRatio, view.padding.top / view.devicePixelRatio, visibleScreenSize.width, visibleScreenSize.height);
+  }
+
+  void _handleContextMenuOnScroll(ScrollNotification notification) {
+    if (_webContextMenuEnabled) {
+      return;
+    }
+    if (!_platformSupportsFadeOnScroll) {
+      _selectionOverlay?.updateForScroll();
+      return;
+    }
+    // When the scroll begins and the toolbar is visible, hide it
+    // until scrolling ends.
+    //
+    // The selection and renderEditable need to be visible within the current
+    // viewport for the toolbar to show when scrolling ends. If they are not
+    // then the toolbar is shown when they are scrolled back into view, unless
+    // invalidated by a change in TextEditingValue.
+    if (notification is ScrollStartNotification) {
+      if (_dataWhenToolbarShowScheduled != null) {
+        return;
+      }
+      final bool toolbarIsVisible = _selectionOverlay != null
+                                  && _selectionOverlay!.toolbarIsVisible
+                                  && !_selectionOverlay!.spellCheckToolbarIsVisible;
+      final Rect? selectionBounds = toolbarIsVisible ? renderEditable.getBoxesForSelection(_value.selection)
+                                           .map((TextBox box) => box.toRect())
+                                           .reduce((Rect result, Rect rect) => result.expandToInclude(rect)) : null;
```

Comment:
reduce throws if the method returns an empty list. Would using the selection handle rects work (and it's only 2 rects instead of however many getBoxesForSelection returns)?

Also I'm a bit curious where we place the toolbar when the text is multiple screens long and you press ctrl-A.

Reply:
Good point, I did not think about the case where the selection is collapsed which would return an empty list. I think in that case we can use the local caret rect. I think using the selection handle rects would work, I started implementing it but it seemed to complicate the implementation some since we are now checking against a List<Rect> instead of a single Rect. I think for now its reasonable to use all the rects from `getBoxesForSelection` since we end up merging them into a single rect anyways.

Reply:
`getBoxesForSelection` can be expensive if there are many runs. But I think that can be addressed later.